src/optimagic/logging/base.py

```diff
@@ -0,0 +1,214 @@
+import io
+import warnings
+from abc import ABC, abstractmethod
+from dataclasses import asdict, fields, is_dataclass
+from typing import Any, Generic, Type, TypeVar
+
+import cloudpickle
+import pandas as pd
+
+from optimagic.exceptions import get_traceback
+from optimagic.typing import DictLikeAccess
+
+InputType = TypeVar("InputType", bound=DictLikeAccess)
+OutputType = TypeVar("OutputType", bound=DictLikeAccess)
+
+
+class _KeyValueStore(Generic[InputType, OutputType], ABC):
+    """Generic abstract base class for a key-value store.
+
+    This class defines the basic interface for key-value stores that support
+    insertion and selection of items based on a primary key.
+
+    Args:
+        input_type: The type of input data that can be stored.
+        output_type: The type of output data that can be retrieved.
+        primary_key: The primary key used to uniquely identify items in the store.
+
+    Raises:
+        ValueError: If input_type or output_type is not a dataclass, or if
+                the primary key is not found in output_type fields.
+
+    """
+
+    def __init__(
+        self,
+        input_type: Type[InputType],
+        output_type: Type[OutputType],
+        primary_key: str,
```

Comment:
The word primary key has been used for a different concept in optimagic. We are getting rid of that concept now but it might still be confusing to some people. However, I don't have a better name and in the database world primary_key is the right name.